src/kernel/bpf/prog.rs

```diff
@@ -61,6 +70,90 @@ impl FromStr for BpfTag {
 }
 
 
+/// A cache for BPF program related information.
+///
+/// The cache allows for convenient look up of BPF program information
+/// based on BPF tag. It is necessary because the kernel does not
+/// provide the means for mapping from tag to program information, but
+/// requires a linear time scan over all active BPF programs.
+///
+/// The cache does the minimal amount of work to establish the mapping
+/// of BPF program information to BPF tag.
+#[derive(Debug, Default)]
+pub struct BpfInfoCache {
+    /// Our mapping from BPF tag to program information.
+    ///
+    /// Program information is reference counted, because it can be
+    /// shared with sub-programs.
+    cache: RefCell<HashMap<BpfTag, Rc<sys::bpf_prog_info>>>,
+}
+
+impl BpfInfoCache {
+    /// Look up BPF program information from the cache.
+    #[cfg_attr(feature = "tracing", crate::log::instrument(level = tracing::Level::TRACE, skip_all, fields(tag, result), err(level = tracing::Level::INFO)))]
+    fn lookup(&self, tag: BpfTag) -> Result<Option<sys::bpf_prog_info>> {
+        let mut cache = self.cache.borrow_mut();
+        if let Some(info) = cache.get(&tag) {
+            return Ok(Some(**info))
+        }
+
+        // If we didn't find the tag inside the cache we have to assume
+        // that our cache is outdated. Given how BPF program information
+        // retrieval works, we have to iterate over all programs, so
+        // clear the cache to remove potentially stale entries.
+        let () = cache.clear();
+
+        let mut found = None;
+        let mut next_prog_id = 0;
+
+        while let Ok(prog_id) = sys::bpf_prog_get_next_id(next_prog_id) {
+            let fd = sys::bpf_prog_get_fd_from_id(prog_id).with_context(|| {
+                format!("failed to retrieve BPF program file descriptor for program {prog_id}")
+            })?;
+
+            let mut info = sys::bpf_prog_info::default();
```

Comment:
be careful with relying on `::default()`. It's best to do an explicit `memset(0)` (Rust's equivalent of it, of course) of the entire struct. Because if there is some padding at the end of the struct, and it will be left non-zero, kernel will reject the struct. This has bitten systemd before (not for bpf_prog_info specifically, it was a different command).

Unless Rust guarantees that `::default()` will zero initialize everything, including padding between the fields. Does it?

Reply:
Padding bytes are explicit in the struct, everything else is a bug.